workspaces/vscode-python/Dockerfile

```diff
@@ -0,0 +1,105 @@
+# noble is the codename for Ubuntu 24.04 LTS
+FROM codercom/code-server:4.90.3-noble
+# Skip code-server:4.91.0-noble and delay updating beyond this until
+# code-server catches up to vscode 1.91.1 because of possible issues on
+# macOS. See: https://github.com/microsoft/vscode/issues/219893
+
+# On PL, we want to standardize on using 1001:1001 for the user. We change the
+# "coder" account's UID and GID here so that fixuid has no work to do later.
+# This speeds up container loading time drastically and avoids timeouts.
+# With Ubuntu-derived base images, this may be a no-op, because recent Ubuntu
+# images have a default "ubuntu" user 1000:1000, and in that case, the
+# "coder" user may already be configured as 1001:1001.
+USER root
+RUN OLD_UID=$(id -u coder) && \
+    OLD_GID=$(id -g coder) && \
+    NEW_UID=1001 && \
+    NEW_GID=1001 && \
+    groupmod -g $NEW_GID coder && \
+    usermod -u $NEW_UID -g $NEW_GID coder && \
+    find /home -user $OLD_UID -execdir chown -h $NEW_UID {} + && \
+    find /home -group $OLD_GID -execdir chgrp -h $NEW_GID {} +
+
+# Remove sudo permissions for default user, then test
+USER root
+RUN find /etc/sudoers.d -type f -not -name README -delete && \
+    sudo -u coder bash -c 'if sudo echo "Test failed: User can sudo" ; then false ; fi'
+
+# Apt installs: Enable manpages with "unminimize", and then install common utilities.
+USER root
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && \
+    yes | unminimize && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* && \
+    find /tmp -not -path /tmp -delete
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+    gosu curl git htop less nano unzip vim wget zip && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* && \
+    find /tmp -not -path /tmp -delete && \
+    # Test the gosu installation:
+    gosu nobody true
+
+# Make sure to build using bash as the shell so that conda/mamba hooks will
+# work during installation.
+SHELL ["/bin/bash", "-c"]
+
+# Install Python and required packages via miniconda.
+# This is a prerequisite for the VS Code extensions in the following steps.
+USER coder
+COPY requirements.txt /
+RUN export arch=`uname -m` \
+ && curl -sfLO "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-${arch}.sh" \
+ && chmod +x "Miniforge3-Linux-${arch}.sh" \
+ && ./"Miniforge3-Linux-${arch}.sh" -b -p /home/coder/conda \
+ # Install conda and mamba hooks for future interactive bash sessions:
+ && /home/coder/conda/bin/mamba init bash \
+ # Activate hooks in the current noninteractive session:
+ && . "/home/coder/conda/etc/profile.d/conda.sh" \
+ && . "/home/coder/conda/etc/profile.d/mamba.sh" \
+ && mamba activate \
+ # Mamba can install pygraphviz without compile errors if we do so first, separately from the rest of the requirements:
+ && mamba install --yes --quiet $(grep pygraphviz /requirements.txt | head -n 1) \
```

Comment:
This feels very fragile and isn't something we've faced with other Python images. Why is this? Have you reported this upstream to the Mamba folks?

Reply:
I took a look at the python grader image and the jupyterlab image. Both install some packages that are required to build the image, but even with the same packages installed in the vscode image the pygraphviz wheel building still fails. I have not been able to identify the source of error yet. 

With that being said, I don't mind using this extra tweaking for pygraphviz at the moment so that the image can be publicly available. I also haven't used pygraphviz so I can't tell whether it is a good idea to just remove the package.

Reply:
@Victorsss-Orz Is the current version in PR here failing to build for you too? (At least, it was building for me, but maybe it's not deterministic.) @firasm any thoughts on how to deal with pygraphviz here, or remove it instead?

Reply:
I can build the image with the current version. 
I've tried removing this line for pygraphviz and install some of the dependencies directly but it failed. I'm fine with using this hack for pygraphviz only, or if people want that removed it would also be ok.

Reply:
I'd prefer removing Pygraphviz then, document it's missing, and create an issue that captures the reason why so it can be handled in a future PR. 

I think Nathan is right, adding these tweaks to the initial image is asking for trouble later and people will copy/paste it and it'll persist forever.

Reply:
OK, we can do that, but since (py)graphviz is in the grader image, there will be some discrepancy. I think that one or more of the Python packages must have had an inaccurate list of dependencies, so where the grader installs the system package for graphviz first instead, it's basically doing the same workaround to sidestep the issue.

Reply:
Yes I think we accept that the first version won't make 100% of everyone 100% happy but perfect is the enemy of good!

@Victorsss-Orz or @echuber2 would either of you like to tackle this? I think that's the last remaining obstacle.

Reply:
Yes I can try to fix this and do the last round of testing by the end of tomorrow (if not today)

Reply:
> Yes I think we accept that the first version won't make 100% of everyone 100% happy but perfect is the enemy of good!

To be honest I don't think I'm the one insisting on perfection. I'm trying to understand the changing criteria to get this merged. Anything is fine with me.

Reply:
Just jumping into to mention that I'm not dependent on this for the start of the semester. (Don't want y'all to rush because of my comment below.) I ended up making my own. I'll be curious to see what the official image looks like in the end though.

(I mostly just skipped `pygraphviz` and installed other things needed for my course.)

Reply:
It's not my intention to be constantly changing requirements, but it seems like every time I come back to look at this PR, something big has changed (new Ubuntu! Mamba! version pinning!) that I'm going to respond to. Both implementation and review here seem to be a moving target.

I want to get this merged too, clearly this image is in demand. I'll try to spend a little time here tomorrow to see if I can't improve things a bit on my own, while also keeping in mind that perfect is indeed the enemy of good. Thanks for your patience and work, everyone.

Reply:
I have a local version without pygraphviz and the rest works just fine. Should I wait until you try things tomorrow to push this?

Reply:
I poked at this a bit. If I also installed `build-essentials graphviz graphviz-dev` at the OS level, pygraphviz was able to build successfully, but at the cost of a larger image size (2.41GB vs 2.23GB when it's installed with Mamba).

I was hopeful that we could use `mamba` to install _all_ packages to make things less idiosyncratic, but some of the package versions we use don't seem to be available via Conda:

```
13.25 Pinned packages:
13.25   - python 3.10.*
13.25 
13.25 
13.25 Could not solve for environment specs
13.25 The following packages are incompatible
13.25 ├─ matplotlib 3.9.0  does not exist (perhaps a typo or a missing channel);
13.25 ├─ openpyxl 3.1.3  does not exist (perhaps a typo or a missing channel);
13.25 └─ typing-extensions 4.12.0  does not exist (perhaps a typo or a missing channel).
```

Given this, I think doing the mamba/pip split is fine. I updated the documentation in 79002ac to better explain why we're doing things like this (in the end, it's not _just_ to "install without compiler errors").

Reply:
@Victorsss-Orz sorry, I missed your note before I pushed and commented. Let's keep `pygraphviz` installed for parity with our other Python images. I'm a lot more comfortable with the state of things now that I better understand how/why this works.

Reply:
I think I had tried throwing Mamba at the entire requirements file and eventually settled on the split when I saw the same problem. Sorry I didn't take more detailed notes about it to justify the split.